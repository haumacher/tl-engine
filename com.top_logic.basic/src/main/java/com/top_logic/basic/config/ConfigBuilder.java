/*
 * SPDX-FileCopyrightText: 2009 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package com.top_logic.basic.config;

import com.top_logic.basic.annotation.FrameworkInternal;

/**
 * Mutable version of a {@link ConfigurationItem} for reflective construction.
 * 
 * <p>
 * Even if the {@link ConfigurationItem} interface does not declare setter
 * methods, its initial values can be assigned through the creation of the
 * {@link ConfigurationItem} instance through a {@link ConfigBuilder}.
 * </p>
 * 
 * @author <a href="mailto:bhu@top-logic.com">Bernhard Haumacher</a>
 */
public interface ConfigBuilder extends ConfigurationItem {

	/**
	 * Sets the value of the given {@link PropertyDescriptor}.
	 * 
	 * <p>
	 * The same property may be initialized only once. The second call to this
	 * method with the same property results in an {@link IllegalStateException}.
	 * </p>
	 * 
	 * @param property
	 *        the descriptor of the property to set.
	 * @param value
	 *        The value to initialize the property with.
	 */
	public void initValue(PropertyDescriptor property, Object value);

	/**
	 * Creates a (potentially) immutable {@link ConfigurationItem} instance from
	 * the builder.
	 * <p>
	 * <b>The caller has to check the {@link InstantiationContext} for errors</b>, as they are not
	 * thrown but reported there. The returned {@link ConfigurationItem} might be in an inconsistent
	 * state if the {@link InstantiationContext#hasErrors()}.
	 * </p>
	 * <p>
	 * Whether the returned {@link ConfigurationItem} instance is immutable
	 * depends on the {@link ConfigurationItem} interface defining setter
	 * methods or not.
	 * </p>
	 * 
	 * @param instantiationContext
	 *        For instantiating inner configurations and for error reporting. <b>Has to be checked
	 *        for errors by the caller.</b>
	 * 
	 * @return The {@link ConfigurationItem} with values initialized from the
	 *         values set to this builder.
	 */
	public ConfigurationItem createConfig(InstantiationContext instantiationContext);

	/**
	 * Initializes the location from which this configuration was read.
	 * 
	 * @param file The configuration file.
	 * @param line The line number of the start tag.
	 * @param column The column number of the start tag. 
	 */
	default void initLocation(String file, int line, int column) {
		initLocation(LocationImpl.location(file, line, column));
	}

	/**
	 * Initializes the location from which this configuration was read.
	 * 
	 * @param location
	 *        The source location.
	 */
	public void initLocation(Location location);

	/**
	 * Disable the pre-instantiation checks, especially whether mandatory properties are set.
	 */
	@FrameworkInternal
	void disableChecks();

	void resetValueSet(PropertyDescriptor property);

}