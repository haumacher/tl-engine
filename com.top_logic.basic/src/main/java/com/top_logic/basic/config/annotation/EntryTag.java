/*
 * SPDX-FileCopyrightText: 2011 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package com.top_logic.basic.config.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import com.top_logic.basic.config.PropertyKind;

/**
 * Annotation of a custom entry name in the serialized form of a typed configuration property of
 * kind {@link PropertyKind#ARRAY}, {@link PropertyKind#LIST} or {@link PropertyKind#MAP}.
 * 
 * @since 5.7.3
 * 
 * @author <a href="mailto:daniel.busche@top-logic.com">Daniel Busche</a>
 */
@TagName("entry-tag")
@Target({ElementType.METHOD, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
public @interface EntryTag {

	/**
	 * The entry tag name of the annotated property.
	 */
	String value();

}

