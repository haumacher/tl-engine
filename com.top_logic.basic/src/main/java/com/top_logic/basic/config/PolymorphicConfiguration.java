/*
 * SPDX-FileCopyrightText: 2009 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package com.top_logic.basic.config;

import com.top_logic.basic.config.annotation.Hidden;
import com.top_logic.basic.config.annotation.Name;
import com.top_logic.basic.config.annotation.ReadOnly;

/**
 * Base interface for {@link ConfigurationItem}s for configuring a polymorphic
 * component hierarchy, where subclasses require specialized configuration.
 * 
 * @see TypedConfiguration Generically instantiating configuration items from
 *      {@link ConfigurationItem} interface declarations.
 * 
 * @author <a href="mailto:bhu@top-logic.com">Bernhard Haumacher</a>
 */
public interface PolymorphicConfiguration<T> extends ConfigurationItem {

	/**
	 * Configuration option for #getImplementationClass()
	 */
	String IMPLEMENTATION_CLASS_NAME = "class";
	
	/**
	 * The corresponding implementation class for this {@link ConfigurationItem}.
	 * 
	 * <p>
	 * In a UI, the implementation class cannot be edited. Instead, the whole configuration item
	 * must be replaced with another implementation class. Otherwise, the properties defined by the
	 * item may be inconsistent with the requirements of the implementation class.
	 * </p>
	 */
	@Name(IMPLEMENTATION_CLASS_NAME)
	@Hidden
	@ReadOnly
	Class<? extends T> getImplementationClass();

	/**
	 * @see #getImplementationClass()
	 */
	void setImplementationClass(Class<? extends T> value);

}
