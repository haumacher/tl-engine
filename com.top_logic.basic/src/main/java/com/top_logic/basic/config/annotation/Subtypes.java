/*
 * SPDX-FileCopyrightText: 2013 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package com.top_logic.basic.config.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import com.top_logic.basic.config.ConfigurationItem;
import com.top_logic.basic.config.PropertyDescriptor;
import com.top_logic.basic.config.PropertyKind;

/**
 * Annotation for a {@link PropertyKind#ARRAY}, {@link PropertyKind#LIST} or
 * {@link PropertyKind#MAP} valued configuration property getter to denote common element types.
 * 
 * @see #value()
 * 
 * @author <a href="mailto:bhu@top-logic.com">Bernhard Haumacher</a>
 */
@Target({ ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
@TagName("sub-types")
public @interface Subtypes {

	/**
	 * Specification of bindings of element tags to common sub-types of the
	 * {@link PropertyDescriptor#getElementType() property element type}.
	 */
	Subtype[] value();

	/**
	 * Whether the annotated {@link Subtype}s are added to/replace the {@link TagName}s of the types
	 * found by type index lookup.
	 */
	boolean adjust() default false;

	/**
	 * Binding of a {@link #type() configuration interface type} to a {@link #tag() tag name}.
	 */
	@TagName("sub-type")
	public @interface Subtype {
		/**
		 * The tag name to associate a certain {@link #type()} with.
		 */
		String tag();

		/**
		 * The configuration interface type to instantiate for a certain element {@link #tag()}.
		 */
		Class<? extends ConfigurationItem> type();
	}

}
