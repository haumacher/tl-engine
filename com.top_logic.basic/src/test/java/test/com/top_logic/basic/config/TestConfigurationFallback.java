/*
 * SPDX-FileCopyrightText: 2011 (c) Business Operation Systems GmbH <info@top-logic.com>
 * 
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-BOS-TopLogic-1.0
 */
package test.com.top_logic.basic.config;

import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import junit.framework.AssertionFailedError;
import junit.framework.Test;
import junit.framework.TestSuite;

import test.com.top_logic.basic.BasicTestCase;
import test.com.top_logic.basic.BasicTestSetup;

import com.top_logic.basic.config.ConfigBuilder;
import com.top_logic.basic.config.ConfigurationDescriptor;
import com.top_logic.basic.config.ConfigurationException;
import com.top_logic.basic.config.ConfigurationItem;
import com.top_logic.basic.config.ConfigurationReader;
import com.top_logic.basic.config.InstantiationContext;
import com.top_logic.basic.config.PolymorphicConfiguration;
import com.top_logic.basic.config.PropertyDescriptor;
import com.top_logic.basic.config.TypedConfiguration;
import com.top_logic.basic.config.annotation.EntryTag;
import com.top_logic.basic.config.annotation.Key;
import com.top_logic.basic.config.annotation.Name;
import com.top_logic.basic.config.annotation.Subtypes;
import com.top_logic.basic.config.annotation.Subtypes.Subtype;
import com.top_logic.basic.config.annotation.defaults.ClassDefault;
import com.top_logic.basic.config.annotation.defaults.IntDefault;
import com.top_logic.basic.config.annotation.defaults.StringDefault;
import com.top_logic.basic.io.character.CharacterContents;

/**
 * The class {@link TestConfigurationFallback} tests fallback to other
 * configurations.
 * 
 * @author <a href="mailto:daniel.busche@top-logic.com">Daniel Busche</a>
 */
@SuppressWarnings("javadoc")
public class TestConfigurationFallback extends AbstractTypedConfigurationTestCase {

	private static Map<String, ConfigurationDescriptor> base() {
		return Collections.singletonMap("base", TypedConfiguration.getConfigurationDescriptor(Base.class));
	}

	public interface A extends ConfigurationItem {


		String A_STRING_VALUE_DEFAULT_VALUE = "aDefault";

		String STRING_VALUE_PROPERTY_NAME = "stringValue";

		@StringDefault(A_STRING_VALUE_DEFAULT_VALUE)
		@Name(STRING_VALUE_PROPERTY_NAME)
		String getStringValue();

		void setStringValue(String stringValue);

		int A_INT_VALUE_DEFAULT_VALUE = 13;
		String INT_VALUE_PROPERTY_NAME = "intValue";

		@IntDefault(A_INT_VALUE_DEFAULT_VALUE)
		@Name(INT_VALUE_PROPERTY_NAME)
		int getIntValue();

		void setIntValue(int intValue);

		B getBConfig();

		@EntryTag("aConfig")
		@Key(STRING_VALUE_PROPERTY_NAME)
		@Subtypes({
			@Subtype(tag = "bConfig", type = B.class),
			@Subtype(tag = "cConfig", type = C.class),
		})
		Map<String, A> getAMapConfigs();

		@EntryTag("aConfig")
		@Subtypes({
			@Subtype(tag = "bConfig", type = B.class),
			@Subtype(tag = "cConfig", type = C.class),
		})
		@Key(STRING_VALUE_PROPERTY_NAME)
		List<A> getAListConfigs();
		
		@Subtypes({})
		PolymorphicConfiguration<? extends Object> getImpl();
	}

	public interface B extends A {
		String B_STRING_VALUE_DEFAULT_VALUE = "bDefault";
		@Override
		@StringDefault(B_STRING_VALUE_DEFAULT_VALUE)
		String getStringValue();

		int B_INT_VALUE_DEFAULT_VALUE = 15;
		@Override
		@IntDefault(B_INT_VALUE_DEFAULT_VALUE)
		int getIntValue();
	}

	/**
	 * Extension of {@link A} which has a {@link Map} of {@link ConfigurationItem}s identified by
	 * the {@link ConfigurationItem#getConfigurationInterface()}
	 * 
	 * @author <a href="mailto:daniel.busche@top-logic.com">Daniel Busche</a>
	 */
	public interface C extends A {
		
		/**
		 * A {@link Map} of {@link ConfigurationItem}s identified by the
		 * {@link ConfigurationItem#getConfigurationInterface()}
		 */
		@Key(CONFIGURATION_INTERFACE_NAME)
		@Subtypes({})
		Map<Class<?>, ConfigurationItem> getConfigItems();

		/**
		 * Simple property not existing in {@link A}
		 */
		Integer getNotInA();
	}
	
	public interface D extends PolymorphicConfiguration<Object> {

		@ClassDefault(Dimpl.class)
		@Override
		public Class<? extends Object> getImplementationClass();
	}

	public static class Dimpl {

	}

	public static class DSpecial extends Dimpl {

	}

	public static class DImplSub {

		public interface Config extends D {
			String getString();
		}

		public DImplSub(InstantiationContext context, Config config) {
		}

	}

	public void testFallback1() throws ConfigurationException {
		final A aConfig = TypedConfiguration.newConfigItem(A.class);
		final B bConfig = newBConfig(aConfig);
		assertEquals(B.B_STRING_VALUE_DEFAULT_VALUE, bConfig.getStringValue());

		String newBValue = "bValue";
		bConfig.setStringValue(newBValue);
		assertEquals(newBValue, bConfig.getStringValue());
	}

	private PropertyDescriptor bProperty(String propertyName) {
		final ConfigurationDescriptor bDescr = TypedConfiguration.getConfigurationDescriptor(B.class);
		final PropertyDescriptor property = bDescr.getProperty(propertyName);
		return property;
	}

	private B newBConfig(final ConfigurationItem fallback) throws ConfigurationException {
		final ConfigBuilder bConfigbuilder = TypedConfiguration.createConfigBuilder(B.class);
		TypedConfiguration.applyFallback(fallback, bConfigbuilder, context);
		B result = (B) bConfigbuilder.createConfig(context);
		// Report errors asap, so they are not hidden by failing assertions.
		context.checkErrors();
		return result;
	}

	public void testFallback2() throws ConfigurationException {
		final A aConfig = TypedConfiguration.newConfigItem(A.class);
		String stringValue = "aValue";
		aConfig.setStringValue(stringValue);
		assertEquals(stringValue, aConfig.getStringValue());

		final B bConfig = newBConfig(aConfig);
		assertEquals(stringValue, bConfig.getStringValue());

		String newBValue = "bValue";
		bConfig.setStringValue(newBValue);
		assertEquals(newBValue, bConfig.getStringValue());

	}

	public void testFallbackPrimitive1() throws ConfigurationException {
		final ConfigurationItem aConfig = TypedConfiguration.newConfigItem(A.class);
		final B bConfig = newBConfig(aConfig);
		assertEquals(B.B_INT_VALUE_DEFAULT_VALUE, bConfig.getIntValue());

		int newBValue = 34;
		bConfig.setIntValue(newBValue);
		assertEquals(newBValue, bConfig.getIntValue());
	}

	public void testFallbackPrimitive2() throws ConfigurationException {
		final A aConfig = TypedConfiguration.newConfigItem(A.class);
		int intValue = 34;
		aConfig.setIntValue(intValue);
		assertEquals(intValue, aConfig.getIntValue());

		final B bConfig = newBConfig(aConfig);
		assertEquals(intValue, bConfig.getIntValue());

		int newBValue = 245;
		bConfig.setIntValue(newBValue);
		assertEquals(newBValue, bConfig.getIntValue());
	}

	public void testInterfaceSpecializationNewPropertyDefaultValue() throws ConfigurationException {
		A a = TypedConfiguration.newConfigItem(A.class);
		ConfigBuilder cBuilder = TypedConfiguration.createConfigBuilder(C.class);
		TypedConfiguration.applyFallback(a, cBuilder, context);

		ConfigurationItem config = cBuilder.createConfig(context);
		// Report errors asap, so they are not hidden by failing assertions.
		context.checkErrors();
		assertEquals(C.class, config.getConfigurationInterface());
		C c = (C) config;
		// Access default value of primitive property which does not exist in fallback item
		assertNull(c.getNotInA());
	}

	public void testInterfaceSpecializationRedeclaredPropertyDefaultValue() throws ConfigurationException {
		A a = TypedConfiguration.newConfigItem(A.class);
		ConfigBuilder bBuilder = TypedConfiguration.createConfigBuilder(B.class);
		TypedConfiguration.applyFallback(a, bBuilder, context);
		ConfigurationItem config = bBuilder.createConfig(context);
		// Report errors asap, so they are not hidden by failing assertions.
		context.checkErrors();
		assertEquals(B.class, config.getConfigurationInterface());
		B b = (B) config;
		// Access default value of primitive property which exist in fallback item but is redeclared
		// in concrete item, i.e. access occurs with a different PropertyDescriptor.
		assertEquals(B.B_STRING_VALUE_DEFAULT_VALUE, b.getStringValue());
	}

	public void testInterfaceSpecializationRedeclaredPropertySetValue() throws ConfigurationException {
		A a = TypedConfiguration.newConfigItem(A.class);
		int intValue = 17;
		assertNotSame(intValue, A.A_INT_VALUE_DEFAULT_VALUE);
		assertNotSame(intValue, B.B_INT_VALUE_DEFAULT_VALUE);
		a.setIntValue(intValue);
		ConfigBuilder bBuilder = TypedConfiguration.createConfigBuilder(B.class);
		TypedConfiguration.applyFallback(a, bBuilder, context);
		ConfigurationItem config = bBuilder.createConfig(context);
		// Report errors asap, so they are not hidden by failing assertions.
		context.checkErrors();
		assertEquals(B.class, config.getConfigurationInterface());
		B b = (B) config;
		assertEquals(intValue, b.getIntValue());
		PropertyDescriptor intProperty = b.descriptor().getProperty(A.INT_VALUE_PROPERTY_NAME);
		assertEquals(intValue, b.value(intProperty));
	}

	public void testListMoveUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "listUpdate.move.expected.xml", "listUpdate.move.xml", "listUpdate.move.increment.xml");
	}

	public void testListMoveStartEndUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName,
			"listUpdate.moveStartEnd.expected.xml",
			"listUpdate.moveStartEnd.xml",
			"listUpdate.moveStartEnd.increment.xml");
	}

	public void testListUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "listUpdate.expected.xml", "listUpdate.xml", "listUpdate.increment.xml");
	}
	
	public void testListAddOrUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "listAddOrUpdate.expected.xml", "listAddOrUpdate.xml", "listAddOrUpdate.increment.xml");
	}
	
	public void testSpecialisationListAddOrUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName,
			"listSpecialisationAddOrUpdate.expected.xml",
			"listSpecialisationAddOrUpdate.xml",
			"listSpecialisationAddOrUpdate.increment.xml");
	}
	
	public void testMapSpecialisationAddOrUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName,
			"mapSpecialisationAddOrUpdate.expected.xml",
			"mapSpecialisationAddOrUpdate.xml",
			"mapSpecialisationAddOrUpdate.increment.xml");
	}

	public void testImplUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "implUpdate.increment.expected.xml",
			"implUpdate.xml", "implUpdate.increment.xml");
	}

	public void testSimpleItemUpdate() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "itemUpdate.increment.expected.xml", "itemUpdate.xml", "itemUpdate.increment.xml");
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "itemUpdate.override.expected.xml", "itemUpdate.xml", "itemUpdate.override.xml");
	}
	
	public void testImplementationClass() throws ConfigurationException, IOException {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "implClass.increment.expected.xml",
			"implClass.xml", "implClass.increment.xml");
	}

	public void testMapUpdate() throws ConfigurationException, IOException {
		HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "mapUpdate.increment.expected.xml", "mapUpdate.xml", "mapUpdate.increment.xml");
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "mapUpdate.override.expected.xml", "mapUpdate.xml", "mapUpdate.override.xml");
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "mapAddOrUpdate.expected.xml", "mapUpdate.xml", "mapAddOrUpdate.increment.xml");
	}

	public interface Base extends ConfigurationItem {
		@Key(Value.KEY)
		@EntryTag(Value.TAG_NAME)
		Collection<Value> getCollection();

		@Key(Value.KEY)
		@EntryTag(Value.TAG_NAME)
		Map<String, Value> getMap();

		@Key(Value.KEY)
		@EntryTag(Value.TAG_NAME)
		List<Value> getList();
	}
	
	public interface Value extends ConfigurationItem {
		String TAG_NAME = "value";

		String KEY = "key";

		@Name(KEY)
		@StringDefault("foobar")
		String getKey();

		int getX();
	}
	
	public void testDefaultMapKey() throws Exception {
		assertEquals(TestConfigurationFallback.class, base(),
			"defaultMapKey.expected.xml", "defaultMapKey.xml", "defaultMapKey.increment.xml");
	}
	
	public void testDefaultListKey() throws Exception {
		assertEquals(TestConfigurationFallback.class, base(),
			"defaultListKey.expected.xml", "defaultListKey.xml", "defaultListKey.increment.xml");
	}

	public void testDefaultCollectionKey() throws Exception {
		assertEquals(TestConfigurationFallback.class, base(),
			"defaultCollectionKey.expected.xml", "defaultCollectionKey.xml", "defaultCollectionKey.increment.xml");
	}

	/**
	 * Tests that the {@link ConfigurationItem#getConfigurationInterface()} can be used as key in a
	 * map valued {@link PropertyDescriptor}.
	 */
	public void testMapByConfigInterface() throws ConfigurationException, IOException {
		HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();
		
		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "mapByInterface.expected.xml",
			"mapByInterface.xml", "mapByInterface.increment.xml");
	}
	
	/**
	 * Tests that in an indexed list the last element can be removed when updating configuration.
	 */
	public void testRemoveLastListElement() throws ConfigurationException, IOException {
		Map<String, ConfigurationDescriptor> globalDescriptorsByLocalName = getDescriptors();

		assertEquals(TestConfigurationFallback.class, globalDescriptorsByLocalName, "listRemoveLast.expected.xml",
			"listRemoveLast.xml", "listRemoveLast.increment.xml", "listRemoveLast.increment2.xml");
	}

	public interface ScenarioTypeItemOwner extends ConfigurationItem {

		ScenarioTypeItem getItem();

	}

	public interface ScenarioTypeItem extends ConfigurationItem {

		String getAlpha();

		String getBeta();
	}

	@SuppressWarnings("unchecked")
	public void testOverrideTrueOnItemOwner() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<ScenarioTypeItemOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <item alpha='a' beta='b' />"
			+ "</ScenarioTypeItemOwner>";
		String incrementXml = XML_DECLARATION
			+ "<ScenarioTypeItemOwner config:override='true' " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <item alpha='x' />"
			+ "</ScenarioTypeItemOwner>";
		ScenarioTypeItemOwner actualItem =
			read(createDescriptorMap(ScenarioTypeItemOwner.class), baseXml, incrementXml);
		assertNotNull(actualItem.getItem());
		assertEquals("x", actualItem.getItem().getAlpha());
		assertEquals("", actualItem.getItem().getBeta());
	}

	@SuppressWarnings("unchecked")
	public void testOverrideTrueOnItemEntry() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<ScenarioTypeItemOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <item alpha='a' beta='b' />"
			+ "</ScenarioTypeItemOwner>";
		String incrementXml = XML_DECLARATION
			+ "<ScenarioTypeItemOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <item config:override='true' alpha='x' />"
			+ "</ScenarioTypeItemOwner>";
		ScenarioTypeItemOwner actualItem =
			read(createDescriptorMap(ScenarioTypeItemOwner.class), baseXml, incrementXml);
		assertNotNull(actualItem.getItem());
		assertEquals("x", actualItem.getItem().getAlpha());
		assertEquals("", actualItem.getItem().getBeta());
	}

	public interface ScenarioTypeMapOwner extends ConfigurationItem {

		@Key(Value.KEY)
		@EntryTag(Value.TAG_NAME)
		Map<String, Value> getMap();

	}

	@SuppressWarnings("unchecked")
	public void testOverrideTrueOnMapOwner() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map>"
			+ "    <value key='first' x='1' />"
			+ "    <value key='second' x='2' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		String incrementXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner config:override='true' " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map>"
			+ "    <value key='first' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		ScenarioTypeMapOwner actualItem = read(createDescriptorMap(ScenarioTypeMapOwner.class), baseXml, incrementXml);
		assertEquals(1, actualItem.getMap().size());
		assertEquals(0, actualItem.getMap().get("first").getX());
	}

	@SuppressWarnings("unchecked")
	public void testOverrideTrueOnMapTag() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map>"
			+ "    <value key='first' x='1' />"
			+ "    <value key='second' x='2' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		String incrementXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map config:override='true'>"
			+ "    <value key='first' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		ScenarioTypeMapOwner actualItem = read(createDescriptorMap(ScenarioTypeMapOwner.class), baseXml, incrementXml);
		assertEquals(1, actualItem.getMap().size());
		assertEquals(0, actualItem.getMap().get("first").getX());
	}

	@SuppressWarnings("unchecked")
	public void testOverrideTrueOnMapEntry() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map>"
			+ "    <value key='first' x='1' />"
			+ "    <value key='second' x='2' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		String incrementXml = XML_DECLARATION
			+ "<ScenarioTypeMapOwner " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <map>"
			+ "    <value config:override='true' key='first' />"
			+ "  </map>"
			+ "</ScenarioTypeMapOwner>";
		ScenarioTypeMapOwner actualItem = read(createDescriptorMap(ScenarioTypeMapOwner.class), baseXml, incrementXml);
		assertEquals(2, actualItem.getMap().size());
		assertEquals(0, actualItem.getMap().get("first").getX());
		assertEquals(2, actualItem.getMap().get("second").getX());
	}

	public void testMapOverrideKeepsConfigInterface() throws ConfigurationException {
		String baseXml = XML_DECLARATION
			+ "<A " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
			+ "  <a-map-configs>"
			+ "    <aConfig config:interface='" + C.class.getName() + "' stringValue='first' />"
			+ "  </a-map-configs>"
			+ "</A>";
		Map<String, ConfigurationDescriptor> descriptorMap = createDescriptorMap(A.class);
		A baseItem = read(descriptorMap, baseXml);
		A firstC = baseItem.getAMapConfigs().get("first");
		assertNotNull(firstC);
		assertTrue(firstC instanceof C);

		String increment = XML_DECLARATION
				+ "<A " + XML_CONFIG_NAMESPACE_DECLARATION + ">"
				+ "  <a-map-configs>"
				+ "    <aConfig stringValue='first' not-in-a='43' />"
				+ "  </a-map-configs>"
				+ "</A>";
		ConfigurationReader reader = new ConfigurationReader(context, descriptorMap);
		reader.setBaseConfig(baseItem);
		reader.setSource(CharacterContents.newContent(increment));
		ConfigurationItem incrementItem;
		try {
			incrementItem = reader.read();
		} catch (AssertionFailedError ex) {
			throw BasicTestCase.fail(
				"Ticket #19377: Configuration interface of fallback item (C.class) not inherited?", ex);
		}

		assertTrue(incrementItem instanceof A);
		A incrementC = ((A) incrementItem).getAMapConfigs().get("first");
		assertNotNull(incrementC);
		assertTrue(incrementC instanceof C);
		assertSame(43, ((C) incrementC).getNotInA());
	}

	@Override
	protected HashMap<String, ConfigurationDescriptor> getDescriptors() {
		final HashMap<String, ConfigurationDescriptor> globalDescriptorsByLocalName =
			new HashMap<>();
		globalDescriptorsByLocalName.put("fallback-test", TypedConfiguration.getConfigurationDescriptor(A.class));
		globalDescriptorsByLocalName.put("fallback-test-c", TypedConfiguration.getConfigurationDescriptor(C.class));
		globalDescriptorsByLocalName.put("fallback-test-d", TypedConfiguration.getConfigurationDescriptor(D.class));
		return globalDescriptorsByLocalName;
	}

	public static Test suite() {
		return BasicTestSetup.createBasicTestSetup(new TestSuite(TestConfigurationFallback.class));
	}

}

