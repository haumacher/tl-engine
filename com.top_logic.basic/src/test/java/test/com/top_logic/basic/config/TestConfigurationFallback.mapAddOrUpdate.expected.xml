<?xml version='1.0' encoding='utf-8'?>
<fallback-test xmlns:config="http://www.top-logic.com/ns/config/6.0">
	<b-config stringValue="oldStringValue">
		<a-map-configs>
			<aConfig
				config:interface="test.com.top_logic.basic.config.TestConfigurationFallback$B"
				stringValue="three" intValue="-1"></aConfig>
			<aConfig
				config:interface="test.com.top_logic.basic.config.TestConfigurationFallback$A"
				stringValue="two" intValue="-2"></aConfig>
			<aConfig
				config:interface="test.com.top_logic.basic.config.TestConfigurationFallback$B"
				intValue="15" stringValue="four" />
			<aConfig
				config:interface="test.com.top_logic.basic.config.TestConfigurationFallback$B"
				intValue="15" stringValue="one"></aConfig>
		</a-map-configs>
	</b-config>
</fallback-test>

